01/implement-most-frequent-k-elements.checkpoint.md

```diff
@@ -0,0 +1,238 @@
+# Most Frequent K Elements
+
+<!-- prettier-ignore-start -->
+### !challenge
+* type: code-snippet
+* language: python3.6
+* id: c371a9d2-49b1-417b-abc4-3765e5cade18
+* title: Most Frequent K Elements
+* points: 3
+### !question
+
+Create a function, `most_frequent_k_elements`, that takes a non-empty array of integers and return the k most frequent elements. (Note: You may assume k is always valid, 1 ≤ k ≤ number of unique elements.)
+
+**Example 1:**
+
+```
+Input: nums = [1,1,1,2,2,3], k = 2
+Output: [1,2]
+```
+
+**Example 2:**
+```
+Input: nums = [1], k = 1
+Output: [1]
+```
+
+### !end-question
+### !placeholder
+
+```python
+def most_frequent_k_elements(num_rows):
+    """
+    A function that returns the first num_rows of Pascal's triangle.
+  
+    Parameters:
+    arr (array): a non-empty array of integers
+    k (int): number of most frequent integers to return 
+  
+    Returns:
+    array[int]: 1D array that that returns the k most frequent elements
+    """
+    pass
+```
+### !end-placeholder
+### !tests
+```python
+import unittest
+from main import *
+
+class TestChallenge(unittest.TestCase):
+    def test_most_k_frequent_elements_nominal(self):
+        # Arrange
+        arr = [1,1,1,2,2,3]
+        k = 2
+
+        # Act
+        result = most_frequents_k_elements(arr, k)
+
+        # Assert
+        self.assertEqual(result, [1,2])
+
+    def test_pascals_big_number(self):
+        # Arrange
+        arr = [1]
+        k = 1 
+
+
+        # Act
+        result = most_frequent_k_elements(arr, k)
+
+        # Act
+        self.assertEqual(result, [1])
+
+    def test_pascals_negative_number_case(self):
+        # Arrange
+        arr = [-1,-1,-1,-2,-5,-5,3,3,3,3,4]
+        k = 3
+
+        # Act
+        result = most_frequent_k_elements(arr, k)
+
+        # Assert
+        self.assertEqual(result, [3,-1,-5])
+
+    def test_pascals_unordered_numbers_case(self):
+        # Arrange
+        arr = [1,2,1,5,1,5,]
+        k = 2
+
+        # Act
+        result = most_frequent_k_elements(arr, k)
+
+        # Assert
+        self.assertEqual(result, [1,5])
+
+
+
+```
+### !end-tests
+### !explanation
+
+An example of a working implementation:
+
+```python
+# Time complexity: O(N + (K - 1)^2) where N is length of the array and K is the number of unique integers.
+# Space complexity: O(N + K) where N is length of the array and K is the number of unique integers.
+def top_k_frequent_elements(arr, k):
```

Comment:
I'd bring over the suggestions I made on the instructor implementation here as well.

Reply:
✅

Reply:
Maybe leave off the complexity here, since this will probably be visible before a student writes up their own complexity analysis. Though if you wanted to include it as the explanations for the time and space complexity (clearly marked as the complexity of the sample implementation), I think that's fine.